llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1131,6 +1191,92 @@ Operation *ReverseOp::getTiledImplementation(OpBuilder &builder,
+  }
+  if (op.getNumOutputs() != 1) {
+    return op.emitOpError("expected one output operand");
+  }
+  auto inputType = op.input().getType().cast<ShapedType>();
+  ArrayRef<int64_t> inputShapes = inputType.getShape();
+  if (inputShapes.size() != 1) {
+    return op.emitError("expected bins rank to be one");
+  }
+  if (!inputType.getElementType().isa<mlir::IntegerType>()) {
+    return op.emitError("expected bins element type to be integer");
+  }
+  // Check if the number of input operands is two, which means weights are also
+  // given as input.
+  if (numInputs == 2) {
+    auto weightsType = op.weights().getType().cast<ShapedType>();
+    ArrayRef<int64_t> weightsShapes = weightsType.getShape();
+    if (weightsShapes.size() != 1) {
+      return op.emitError("expected weights rank to be one");
+    }
+    if (inputShapes[0] != weightsShapes[0]) {
+      return op.emitOpError("incompatible bins and weights shapes");
+    }
+    auto outputType = op.output().getType().cast<ShapedType>();
+    // If weights are given as input, the output must be floating-point type.
+    if (!outputType.getElementType().isa<mlir::FloatType>())
+      return op.emitError("expected output element type to be float");
+  }
+  return success();
+}
+
+SmallVector<Range> BinCountOp::getIterationDomain(OpBuilder &builder) {
+  int64_t operandRank = getOperandRank();
+  SmallVector<Range> loopBounds(operandRank);
+  Location loc = getLoc();
+  Value zero = builder.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = builder.create<arith::ConstantIndexOp>(loc, 1);
+  Value source = input();
+  for (auto dim : llvm::seq<int64_t>(0, operandRank)) {
+    loopBounds[dim].offset = zero;
+    loopBounds[dim].size = getDimValue(builder, loc, source, dim);
+    loopBounds[dim].stride = one;
+  }
+  return loopBounds;
+}
+
+SmallVector<StringRef> BinCountOp::getLoopIteratorTypes() {
+  SmallVector<StringRef> iteratorTypes(getOperandRank(),
+                                       getParallelIteratorTypeName());
+  return iteratorTypes;
+}
+
+// Op description is given at
+// https://pytorch.org/docs/stable/generated/torch.bincount.html.
+LogicalResult BinCountOp::generateScalarImplementation(OpBuilder &b,
+                                                       Location loc,
+                                                       ValueRange ivs) {
+  Value input_ = input();
+  Value output_ = output();
+  Type outputType = output_.getType().cast<ShapedType>().getElementType();
+  Value one = b.create<arith::ConstantOp>(loc, b.getIntegerAttr(outputType, 1));
+  Value bin = b.create<memref::LoadOp>(loc, input_, ivs.front());
+  bin = castIntToIndex(b, loc, bin);
+  Value count = b.create<memref::LoadOp>(loc, output_, bin);
+  if (getNumInputs() == 2) {
+    Value weights_ = weights();
+    Value weight = b.create<memref::LoadOp>(loc, weights_, ivs.front());
+    weight = convertScalarToDtype(b, loc, weight, outputType);
```

Comment:
why do you need to do convertScalarToDtype? If we require weight and output to be the same type does this go away?

Reply:
This comment is linked to your previous comment, and I have answered that.

Reply:
Sounds good!